File: port/java/src/com/gams/variables/Sensor.java
/*********************************************************************
 * Copyright (c) 2013-2015 Carnegie Mellon University. All Rights Reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following acknowledgments and disclaimers.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 * this list of conditions and the following disclaimer in the documentation
 * and/or other materials provided with the distribution.
 * 
 * 3. The names "Carnegie Mellon University," "SEI" and/or
 * "Software Engineering Institute" shall not be used to endorse or promote
 * products derived from this software without prior written permission. For
 * written permission, please contact permission@sei.cmu.edu.
 * 
 * 4. Products derived from this software may not be called "SEI" nor may "SEI"
 * appear in their names without prior written permission of
 * permission@sei.cmu.edu.
 *
 * 5. Redistributions of any form whatsoever must retain the following
 * acknowledgment:
 *
 * This material is based upon work funded and supported by the Department of
 * Defense under Contract No. FA8721-05-C-0003 with Carnegie Mellon University
 * for the operation of the Software Engineering Institute, a federally funded
 * research and development center. Any opinions, findings and conclusions or
 * recommendations expressed in this material are those of the author(s) and
 * do not necessarily reflect the views of the United States Department of
 * Defense.
 * 
 * NO WARRANTY. THIS CARNEGIE MELLON UNIVERSITY AND SOFTWARE ENGINEERING
 * INSTITUTE MATERIAL IS FURNISHED ON AN "AS-IS" BASIS. CARNEGIE MELLON
 * UNIVERSITY MAKES NO WARRANTIES OF ANY KIND, EITHER EXPRESSED OR IMPLIED,
 * AS TO ANY MATTER INCLUDING, BUT NOT LIMITED TO, WARRANTY OF FITNESS FOR
 * PURPOSE OR MERCHANTABILITY, EXCLUSIVITY, OR RESULTS OBTAINED FROM USE OF THE
 * MATERIAL. CARNEGIE MELLON UNIVERSITY DOES NOT MAKE ANY WARRANTY OF ANY KIND
 * WITH RESPECT TO FREEDOM FROM PATENT, TRADEMARK, OR COPYRIGHT INFRINGEMENT.
 * 
 * This material has been approved for public release and unlimited
 * distribution.
 * 
 * @author James Edmondson <jedmondson@gmail.com>
 *********************************************************************/
package com.gams.variables;

import java.util.HashSet;

import com.gams.GamsJNI;
import com.madara.KnowledgeBase;
import com.madara.containers.String;
import com.madara.containers.Vector;
import com.gams.utility.Position;
import com.gams.utility.GpsPosition;

public class Sensor extends GamsJNI
{	
  private native long jni_Sensor();
  private native long jni_Sensor(long cptr);
  private static native void jni_freeSensor(long cptr);
  private native java.lang.String jni_getName(long cptr);
  private native void jni_init(long cptr, long type, long kb, java.lang.String name, double range);
  private native java.lang.String jni_toString(long cptr);
  private native double jni_getPositionValue(long cptr, long pos);
  private native double jni_getGpsValue(long cptr, long pos);
  private native long jni_getOrigin(long cptr);
  private native double jni_getRange(long cptr);
  private native void jni_setOrigin(long cptr, long origin);
  private native void jni_setRange(long cptr, double range);
  private native void jni_setPositionValue(long cptr, long pos, double value);
  private native void jni_setGpsValue(long cptr, long pos, double value);
  private native long jni_getGpsFromIndex(long cptr, long index);
  private native long jni_getIndexFromGps(long cptr, long position);
  private native double jni_getDiscretization(long cptr);
  private native long[] jni_discretizeRegion(long cptr, long region);
  private native long[] jni_discretizeSearchArea(long cptr, long area);

  private boolean manageMemory = true;

  /**
   * Default constructor
   **/
  public Sensor()
  {
    setCPtr(jni_Sensor());
  }

  /**
   * Copy constructor
   * @param input the sensor value to copy
   **/
  public Sensor(Sensor input)
  {
    setCPtr(jni_Sensor(input.getCPtr()));
  }

  /**
   * Gets the sensor range in meters.
   * @return range in meters
   **/
  public GpsPosition getOrigin()
  {
    return GpsPosition.fromPointer (jni_getOrigin(getCPtr()));
  }
  
  /**
   * Gets the sensor range in meters.
   * @return range in meters
   **/
  public double getRange()
  {
    return jni_getRange(getCPtr());
  }
  
  /**
   * Gets a GPS coordinate from an index
   * @param index the cartesian position
   * @return the position at the specified index
   **/
  public GpsPosition getGpsFromIndex(Position index)
  {
    return GpsPosition.fromPointer(
      jni_getGpsFromIndex(getCPtr(),index.getCPtr()));
  }
    
  /**
   * Gets a GPS coordinate from an index
   * @param coord coordinate to convert to an index 
   * @return the position at the specified index
   **/
  public Position getIndexFromGps(GpsPosition coord)
  {
    return Position.fromPointer(
      jni_getIndexFromGps(getCPtr(),coord.getCPtr()));
  }
  
  /**
   * Gets the value stored at a particular location
   * @param  position  the location to check
   * @return value at the location
   **/
  public double getValue(GpsPosition position)
  {
    return jni_getGpsValue(getCPtr(), position.getCPtr());
  }
    
  /**
   * Gets the value stored at a particular location
   * @param  position  the location to check
   * @return value at the location
   **/
  public double getValue(Position position)
  {
    return jni_getPositionValue(getCPtr(), position.getCPtr());
  }
       
  /**
   * Gets the discretization value for the sensor
   * @return the discretization value for the sensor
   **/
  public double getDiscretization()
  {
    return jni_getDiscretization(getCPtr());
  }
        
  /**
   * Sets the range of the sensor
   * @param  range  the range of the sensor in meters
   **/
  public void setRange(double range)
  {
    jni_setRange(getCPtr(), range);
  }
          
  /**
   * Sets the origin for coverage information and discretization
   * @param  origin  the origin for coverage information
   **/
  public void setOrigin(GpsPosition origin)
  {
    jni_setOrigin(getCPtr(), origin.getCPtr());
  }
         
  /**
   * Sets a value at a location
   * @param  location  the position the value will be at
   * @param  value     the value to set
   **/
  public void setValue(Position location, double value)
  {
    jni_setPositionValue(getCPtr(), location.getCPtr(), value);
  }
          
  /**
   * Sets a value at a location
   * @param  location  the position the value will be at
   * @param  value     the value to set
   **/
  public void setValue(GpsPosition location, double value)
  {
    jni_setPositionValue(getCPtr(), location.getCPtr(), value);
  }
            
  /**
   * Discretizes the region into individual positions
   * @param  region  the region to discretize
   * @return individual positions within the region
   **/
  public HashSet<Position> discretize(com.gams.utility.Region region)
  {
    long[] indices = jni_discretizeRegion(getCPtr(), region.getCPtr());

    HashSet<Position> hash = new HashSet<Position> ();
    for (int i = 0; i < indices.length; ++i)
      hash.add(Position.fromPointer(indices[i]));

    return hash;
  }
              
  /**
   * Discretizes the area into individual positions
   * @param  area  the area to discretize
   * @return individual positions within the area
   **/
  public HashSet<Position> discretize(com.gams.utility.SearchArea area)
  {
    long[] indices = jni_discretizeSearchArea(getCPtr(), area.getCPtr());

    HashSet<Position> hash = new HashSet<Position> ();
    for (int i = 0; i < indices.length; ++i)
      hash.add(Position.fromPointer(indices[i]));

    return hash;
  }
  
  /**
   * Creates a java object instance from a C/C++ pointer
   *
   * @param cptr C pointer to the object
   * @return a new java instance of the underlying pointer
   */
  public static Sensor fromPointer(long cptr)
  {
    Sensor ret = new Sensor();
    ret.manageMemory = true;
    ret.setCPtr(cptr);
    return ret;
  }

  /**
   * Creates a java object instance from a C/C++ pointer
   *
   * @param cptr C pointer to the object
   * @param shouldManage  if true, manage the pointer
   * @return a new java instance of the underlying pointer
   */
  public static Sensor fromPointer(long cptr, boolean shouldManage)
  {
    Sensor ret = new Sensor();
    ret.manageMemory=shouldManage;
    ret.setCPtr(cptr);
    return ret;
  }

  /**
   * Gets the name of the variable
   *
   * @return  name of the variable within the context
   */
  public java.lang.String getName()
  {
    return jni_getName(getCPtr());
  }

  /**
   * Sets the name and knowledge base being referred to
   *
   * @param  kb      the knowledge base that contains the name
   * @param  name    the variable name
   * @param  range   the range of the sensor in meters or appropriate unit
   */
  public void init(KnowledgeBase kb, java.lang.String name, double range)
  {
    jni_init(getCPtr(), 0, kb.getCPtr (), name, range);
  }

  /**
   * Converts the value to a string
   *
   * @return current string value
   */
  public java.lang.String toString()
  {
    return jni_toString(getCPtr());
  }

  /**
   * Deletes the C instantiation. To prevent memory leaks, this <b>must</b> be
   * called before an instance gets garbage collected
   */
  public void free()
  {
    if (manageMemory)
    {
      jni_freeSensor(getCPtr());
    }
  }
  
  /**
   * Cleans up underlying C resources
   * @throws Throwable necessary for override but unused
   */
  @Override
  protected void finalize() throws Throwable
  {
    try {
      free();
    } catch (Throwable t) {
      throw t;
    } finally {
      super.finalize();
    }
  }
}

File: src/gams/variables/Sensor.cpp
/**
 * Copyright (c) 2014 Carnegie Mellon University. All Rights Reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 * 
 * 1. Redistributions of source code must retain the above copyright notice,
 *    this list of conditions and the following acknowledgments and disclaimers.
 * 
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 * 
 * 3. The names "Carnegie Mellon University," "SEI" and/or "Software
 *    Engineering Institute" shall not be used to endorse or promote products
 *    derived from this software without prior written permission. For written
 *    permission, please contact permission@sei.cmu.edu.
 * 
 * 4. Products derived from this software may not be called "SEI" nor may "SEI"
 *    appear in their names without prior written permission of
 *    permission@sei.cmu.edu.
 * 
 * 5. Redistributions of any form whatsoever must retain the following
 *    acknowledgment:
 * 
 *      This material is based upon work funded and supported by the Department
 *      of Defense under Contract No. FA8721-05-C-0003 with Carnegie Mellon
 *      University for the operation of the Software Engineering Institute, a
 *      federally funded research and development center. Any opinions,
 *      findings and conclusions or recommendations expressed in this material
 *      are those of the author(s) and do not necessarily reflect the views of
 *      the United States Department of Defense.
 * 
 *      NO WARRANTY. THIS CARNEGIE MELLON UNIVERSITY AND SOFTWARE ENGINEERING
 *      INSTITUTE MATERIAL IS FURNISHED ON AN "AS-IS" BASIS. CARNEGIE MELLON
 *      UNIVERSITY MAKES NO WARRANTIES OF ANY KIND, EITHER EXPRESSED OR
 *      IMPLIED, AS TO ANY MATTER INCLUDING, BUT NOT LIMITED TO, WARRANTY OF
 *      FITNESS FOR PURPOSE OR MERCHANTABILITY, EXCLUSIVITY, OR RESULTS
 *      OBTAINED FROM USE OF THE MATERIAL. CARNEGIE MELLON UNIVERSITY DOES
 *      NOT MAKE ANY WARRANTY OF ANY KIND WITH RESPECT TO FREEDOM FROM PATENT,
 *      TRADEMARK, OR COPYRIGHT INFRINGEMENT.
 * 
 *      This material has been approved for public release and unlimited
 *      distribution.
 **/

#include "gams/variables/Sensor.h"

#include <float.h>
#include <sstream>
#include <vector>
#include <string>
#include <cmath>

using std::string;
using std::stringstream;
using std::vector;

typedef  madara::knowledge::KnowledgeRecord::Integer  Integer;

gams::variables::Sensor::Sensor () :
  knowledge_ (0), name_ ("")
{
}

gams::variables::Sensor::Sensor (const string & name,
  madara::knowledge::KnowledgeBase * knowledge,
  const double & range, const pose::Position & origin) :
  knowledge_ (knowledge), name_ (name)
{
  init_vars ();

  if (range_ == 0.0 && range != 0.0)
    range_ = range;
  if (origin.latitude () != DBL_MAX)
    origin.to_container (origin_);
}

gams::variables::Sensor::~Sensor ()
{
}

void
gams::variables::Sensor::operator= (const Sensor & rhs)
{
  if (this != &rhs)
  {
    this->range_ = rhs.range_;
    this->value_ = rhs.value_;
    this->origin_ = rhs.origin_;
    this->knowledge_ = rhs.knowledge_;
    this->name_ = rhs.name_;
  }
}

set<gams::pose::Position>
gams::variables::Sensor::discretize (
  const pose::Region & region)
{
  set<pose::Position> ret_val;

  // find northern most point
  pose::Position northern = region.vertices[0];
  for (size_t i = 1; i < region.vertices.size (); ++i)
    if (northern.latitude () < region.vertices[i].latitude ())
      northern = region.vertices[i];
  const int max_x = (int)get_index_from_gps (northern).x();

  // find southern most point
  pose::Position southern = region.vertices[0];
  for (size_t i = 1; i < region.vertices.size (); ++i)
    if (southern.latitude () > region.vertices[i].latitude ())
      southern = region.vertices[i];
  const int min_x = (int)get_index_from_gps (southern).x();

  // find west most point 
  pose::Position start;
  start.longitude (DBL_MAX);
  for (size_t i = 0; i < region.vertices.size (); ++i)
    if (start.longitude () > region.vertices[i].longitude ())
      start = region.vertices[i];

  // find valid corresponding position
  pose::Position start_index = get_index_from_gps (start);
  if (!region.contains (get_gps_from_index (start_index)))
  {
    start_index.y(start_index.y() + 1); // go one east...
    pose::Position check = start_index;
    while ((!region.contains (get_gps_from_index (check))) &&
      check.x() <= max_x)
    {
      check.x(check.x() + 1); // ... and start looking north for position in region
    }

    // if we still haven't found a good position...
    if (!region.contains (get_gps_from_index (check)))
    {
      // ...start looking south
      check = start_index;
      while ((!region.contains (get_gps_from_index (check))) &&
        check.x() >= min_x)
      {
        check.x(check.x() - 1);
      }
    }

    // by now we must have found a starting position, so update start_index
    start_index = check;
  }

  // find east most point
  pose::Position eastern = region.vertices[0];
  for (size_t i = 1; i < region.vertices.size (); ++i)
    if (eastern.longitude () < region.vertices[i].longitude ())
      eastern = region.vertices[i];
  const int max_y = (int)get_index_from_gps (eastern).y();

  // move east each iteration
  while (start_index.y() < max_y)
  {
    // check north
    for (pose::Position pos = start_index; pos.x() <= max_x; pos.x(pos.x() + 1))
      if (region.contains (get_gps_from_index (pos)))
        ret_val.insert (pos);
  
    // check south
    for (pose::Position pos = start_index; pos.x() >= min_x; pos.x(pos.x() - 1))
      if (region.contains (get_gps_from_index (pos)))
        ret_val.insert (pos);

    start_index.y(start_index.y() + 1);
  }

  return ret_val;
}

set<gams::pose::Position>
gams::variables::Sensor::discretize (
  const pose::SearchArea & search)
{
  set<pose::Position> ret_val;
  const vector<pose::PrioritizedRegion>& regions = search.get_regions ();
  for (size_t i = 0; i < regions.size (); ++i)
  {
    set<pose::Position> to_add = discretize (regions[i]);
    ret_val.insert (to_add.begin (), to_add.end ());
  }
  return ret_val;
}

double
gams::variables::Sensor::get_discretization () const
{
  return sqrt (2.0 * pow(get_range (), 2.0));
}

void
gams::variables::Sensor::regenerate_local_frame ()
{
  local_frame_ = pose::ReferenceFrame(pose::Cartesian, get_origin());
}

gams::pose::Position
gams::variables::Sensor::get_gps_from_index (
  const pose::Position & idx)
{
  const double discretize = get_discretization ();

  regenerate_local_frame();

  pose::Position meters (local_frame_,
    int(idx.x()) * discretize, int(idx.y()) * discretize, int(idx.z()));
  pose::Position ret = meters.transform_to(pose::gps_frame());
  return ret;
}

gams::pose::Position
gams::variables::Sensor::get_index_from_gps (
  const pose::Position & pos)
{
  regenerate_local_frame();

  pose::Position idx = pos.transform_to (local_frame_);

  const double discretize = get_discretization ();
  idx.x((int)((idx.x() + discretize / 2) / discretize));
  idx.y((int)((idx.y() + discretize / 2) / discretize));

  return idx;
}

string
gams::variables::Sensor::get_name () const
{
  return name_;
}

gams::pose::Position
gams::variables::Sensor::get_origin ()
{
  pose::Position origin(pose::gps_frame());
  origin.from_container (origin_);
  return origin;
}

double
gams::variables::Sensor::get_range () const
{
  return range_.to_double ();
}

double
gams::variables::Sensor::get_value (const pose::Position & pos)
{
  return value_[index_pos_to_index (get_index_from_gps (pos))].to_double ();
}

void
gams::variables::Sensor::set_origin (const pose::Position & origin)
{
  origin.to_container (origin_);
}

void
gams::variables::Sensor::set_range (const double & range)
{
  range_ = range;
}

void
gams::variables::Sensor::set_value (const pose::Position & pos,
  const double & val,
  const madara::knowledge::KnowledgeUpdateSettings & settings)
{
  string idx = index_pos_to_index (get_index_from_gps(pos));
  value_.set (idx, val, settings);
}

string
gams::variables::Sensor::index_pos_to_index (
  const pose::Position & pos) const
{
  stringstream buffer;
  buffer << (int)(pos.x()) << "x" << (int)(pos.y());

  return buffer.str ();
}

void
gams::variables::Sensor::init_vars ()
{
  // sensor information is prefixed by sensor.<name_>
  std::string prefix ("sensor");
  prefix += ".";
  prefix += name_;

  // initialize the variable containers
  range_.set_name (prefix + ".range", *knowledge_);
  value_.set_name (prefix + ".covered", *knowledge_);
  origin_.set_name (prefix + ".origin", *knowledge_);
}

void
gams::variables::Sensor::init_vars (const string & name,
  madara::knowledge::KnowledgeBase * knowledge,
  const double & range, const pose::Position & origin)
{
  name_ = name;
  knowledge_ = knowledge;

  /**
   * We only want to update range if it has not yet been set. This could result 
   * in inconsistencies if multiple agents try to set different ranges. In the 
   * best case, there is a round of sensor values that are inconsistent. 
   * Depending on the actual data that is being collected, this might not be an
   * issue.
   */
  if (range_ != 0.0)
    range_ = range;

  /**
   * Similar to range, origin could be set incorrectly by multiple agents. The
   * best option would be to have the controller set it when initializing the
   * system.
   **/
  pose::Position cur_origin(pose::gps_frame());
  cur_origin.from_container (origin_);
  if (cur_origin.latitude () != 0.0 && origin.latitude () != DBL_MAX)
    origin.to_container (origin_);

  init_vars ();
}
